Logging and Exception Management Assignment

anubhab-ti opened the pull request.

anubhab-ti committed:
Added exceptions to routers/lead_conversion.py

anubhab-ti committed:
Added exceptions to routers/submit_lead.py

Removed accidental auto-import from routers/lead_conversion.py

anubhab-ti committed:
Added exceptions to routers/three_pl.py

anubhab-ti committed:
Added exception to database/db_helper.py

anubhab-ti committed:
Setup global logging configuration in main.py

anubhab-ti committed:
Added logging to database/db_helper.py

anubhab-ti committed:
Added logging and exception handling to utils/adf.py

anubhab-ti committed:
Added logging and exception handling to routers/lead_conversion.py

anubhab-ti committed:
Added logging and exception handling to routers/submit_lead.py

anubhab-ti committed:
Added logging and exception handling to routers/three_pl.py

anubhab-ti committed:
Added an exception and logging to services/enrich_lead.py

anubhab-ti committed:
Added logging and exception handling to services/verify_phone_and_email.py

naman-TI commented on fast_api_als/services/verify_phone_and_email.py:
```diff
@@ -38,14 +43,30 @@ async def verify_phone_and_email(email: str, phone_number: str) -> bool:
     phone_valid = False
     data = {}
 
+    logger.info(f'Email validation URL: {email_validation_url}')
```
These logs are not needed at the 'info' level

anubhab-ti replied:
I dropped them to debug level.

naman-TI commented on fast_api_als/services/verify_phone_and_email.py:
```diff
@@ -38,14 +43,30 @@ async def verify_phone_and_email(email: str, phone_number: str) -> bool:
         call_validation_service(phone_validation_url, "phone", phone_number, data),
     )
+    end_time = int(time.time() * 1000.0)
+    logger.info(f'Total time taken by validation service: {end_time - start_time}')
```
no unit for time?

anubhab-ti replied:
I added 'ms' to the end of the message.

naman-TI commented on fast_api_als/utils/adf.py:
```diff
@@ -39,11 +39,15 @@ def is_nan(x):
+        obj = xmltodict.parse(adf_xml)
+        return obj
+    except Exception as e:
+        logger.info(f'xml parse error {e} for adf_xml: {adf_xml}')
```
incorrect log level, and no exception is raised here

anubhab-ti replied:
Do you mean the code in the ```try``` block does not throw any exception? I will fix the log level to error.

naman-TI replied:
look up handling exceptions and what 'raising' an exception means

naman-TI commented on fast_api_als/services/enrich_lead.py:
```diff
@@ -9,7 +9,10 @@
-    pass
\ No newline at end of file
+    if adf_json is None or adf_json == {}:
+        raise ValueError
```
there can be many more exceptions handled here!

naman-TI commented on fast_api_als/routers/lead_conversion.py:
```diff
@@ -42,29 +44,35 @@ def get_quicksight_data(lead_uuid, item):
+    try:
+        body = await file.body()
+        body = json.loads(str(body, 'utf-8'))
+    except Exception as e:
```
the exception caught is too generic

naman-TI commented on fast_api_als/routers/lead_conversion.py:
```diff
@@ -42,29 +44,35 @@ def get_quicksight_data(lead_uuid, item):
     if role != "OEM":
-        # throw proper HTTPException
-        pass
+        raise HTTPException(status_code=401, detail="User with role not OEM is not authorised to submit")
```
logs are also expected here

naman-TI commented on fast_api_als/routers/lead_conversion.py:
```diff
@@ -42,29 +44,35 @@ def get_quicksight_data(lead_uuid, item):
     else:
-        # throw proper HTTPException
-        pass
+        raise HTTPException(status_code=500, detail='Lead Conversion failed')
```
logs are also expected here!

naman-TI commented on fast_api_als/routers/submit_lead.py:
```diff
@@ -31,19 +31,32 @@
+        end_time = int(time.time() * 1000.0)
+        time_taken = end_time - start_time
+        logger.info(f'Time taken to load file body is: {time_taken}')
+    except Exception as e:
```
The exception can only take place is one line, not this whole block.

naman-TI commented on fast_api_als/routers/submit_lead.py:
```diff
@@ -70,6 +84,7 @@ async def submit(file: Request, apikey: APIKey = Depends(get_api_key)):
     if not validation_check:
         item, path = create_quicksight_data(obj['adf']['prospect'], lead_hash, 'REJECTED', validation_code, {})
         s3_helper_client.put_file(item, path)
+        logger.info('Submit failed due to validation error')
```
A better log message is expected here

anubhab-ti committed:
Dropped logging level of some logs to debug in services/verify_phone_and_email

Added unit validation service time log

naman-TI commented on fast_api_als/routers/three_pl.py:
```diff
@@ -29,15 +37,22 @@ async def reset_authkey(request: Request, token: str = Depends(get_token)):
+        body = json.loads(body)
+        logger.info('Body of request parsed successfully')
+    except Exception as e:
+        logger.error(f'Parsing of body failed due to {e}')
```
what is the specific type of exception that is caught here and the similar instances in the same file?

naman-TI commented on fast_api_als/main.py:
```diff
@@ -36,3 +37,12 @@ def ping():
+def main():
+    # setting up global logging configuration with threshold level = INFO and truncate on restart
+    FORMAT = '%(name)s %(asctime)s %(levelname)s:%(message)s'
+    logging.basicConfig(format=FORMAT, level=logging.INFO, filename='fast_api_als.log', filemode='w')
```
please add .log in the .gitignore file as well. If the .gitignore is not created, read up on it

naman-TI reviewed (changes requested):
Please address the mentioned comments and also try to include better exception handling and logs in the rest of the repo.

anubhab-ti committed:
Raised more exceptions in services/enrich_lead.py

anubhab-ti committed:
Fixed log level and raised ValueError in utils/adf.py

anubhab-ti committed:
Added except blocks to catch JSONDecodeError raise by json.loads() call

anubhab-ti committed:
Logged context data before throwing HTTPException in routers/lead_conversion.py

anubhab-ti committed:
Removed time measuring statements to outside of try block in routers/submit_lead.py

Added except block to catch specific httpx.RequestError in verify_phone_and_email.py

anubhab-ti committed:
Improved log message for validation error in routers/submit_lead.py

anubhab-ti committed:
Added .gitignore and added log file to .gitignore
